File: TypeScriptCodes/src/communication/SecureProtocol.ts
import WebSocket from 'ws';
import { QuantumEncryption } from './QuantumEncryption';
import { ErrorHandler, ErrorSeverity } from '../utils/ErrorHandling';
import { ConfigManager } from '../utils/ConfigManager';
import { Message } from './Protocol';

export class SecureProtocol {
    private quantumEnc: QuantumEncryption;
    private sessionKeys: Map<string, CryptoKey>;
    private encryptionKeys: Map<string, CryptoKey>;
    private config: ConfigManager;

    constructor() {
        this.quantumEnc = new QuantumEncryption();
        this.sessionKeys = new Map();
        this.encryptionKeys = new Map();
        this.config = ConfigManager.getInstance();
    }

    async establishSecureChannel(websocket: WebSocket): Promise<string> {
        try {
            // Generate quantum key
            const quantumKey = await this.quantumEnc.generateQuantumKey();

            // Generate session key
            const sessionKey = await this.generateSessionKey();
            const encryptionKey = await this.deriveEncryptionKey(sessionKey);

            // Encrypt session key with quantum key
            const encryptedKey = await this.quantumEnc.encryptMessage(
                this.arrayBufferToBase64(await crypto.subtle.exportKey('raw', sessionKey)),
                quantumKey
            );

            // Send encrypted session key
            websocket.send(JSON.stringify({
                type: 'key_exchange',
                key: this.arrayBufferToBase64(encryptedKey)
            }));

            // Store session information
            const sessionId = this.generateSessionId();
            this.sessionKeys.set(sessionId, sessionKey);
            this.encryptionKeys.set(sessionId, encryptionKey);

            return sessionId;
        } catch (error) {
            throw ErrorHandler.createError(
                `Secure channel establishment failed: ${error}`,
                ErrorSeverity.ERROR,
                'SECURE_PROTOCOL'
            );
        }
    }

    async encryptMessage(sessionId: string, message: Message): Promise<ArrayBuffer> {
        try {
            const key = this.encryptionKeys.get(sessionId);
            if (!key) {
                throw new Error('Invalid session ID');
            }

            const data = new TextEncoder().encode(JSON.stringify(message.toJSON()));
            const iv = crypto.getRandomValues(new Uint8Array(12));
            const encrypted = await crypto.subtle.encrypt(
                { name: 'AES-GCM', iv },
                key,
                data
            );

            // Combine IV and encrypted data
            const result = new Uint8Array(iv.length + encrypted.byteLength);
            result.set(iv);
            result.set(new Uint8Array(encrypted), iv.length);

            return result.buffer;
        } catch (error) {
            throw ErrorHandler.createError(
                `Message encryption failed: ${error}`,
                ErrorSeverity.ERROR,
                'SECURE_PROTOCOL'
            );
        }
    }

    async decryptMessage(sessionId: string, encrypted: ArrayBuffer): Promise<Message> {
        try {
            const key = this.encryptionKeys.get(sessionId);
            if (!key) {
                throw new Error('Invalid session ID');
            }

            // Extract IV and encrypted data
            const data = new Uint8Array(encrypted);
            const iv = data.slice(0, 12);
            const encryptedData = data.slice(12);

            const decrypted = await crypto.subtle.decrypt(
                { name: 'AES-GCM', iv },
                key,
                encryptedData
            );

            const messageData = JSON.parse(new TextDecoder().decode(decrypted));
            return Message.fromJSON(messageData);
        } catch (error) {
            throw ErrorHandler.createError(
                `Message decryption failed: ${error}`,
                ErrorSeverity.ERROR,
                'SECURE_PROTOCOL'
            );
        }
    }

    closeSession(sessionId: string): void {
        this.sessionKeys.delete(sessionId);
        this.encryptionKeys.delete(sessionId);
    }

    private async generateSessionKey(): Promise<CryptoKey> {
        return await crypto.subtle.generateKey(
            { name: 'AES-GCM', length: 256 },
            true,
            ['encrypt', 'decrypt']
        );
    }

    private async deriveEncryptionKey(sessionKey: CryptoKey): Promise<CryptoKey> {
        const salt = crypto.getRandomValues(new Uint8Array(16));
        return await crypto.subtle.deriveKey(
            {
                name: 'PBKDF2',
                salt,
                iterations: 100000,
                hash: 'SHA-256'
            },
            sessionKey,
            { name: 'AES-GCM', length: 256 },
            true,
            ['encrypt', 'decrypt']
        );
    }

    private generateSessionId(): string {
        return crypto.randomUUID();
    }

    private arrayBufferToBase64(buffer: ArrayBuffer): string {
        return Buffer.from(buffer).toString('base64');
    }

    private base64ToArrayBuffer(base64: string): ArrayBuffer {
        return Buffer.from(base64, 'base64');
    }
} 

File: PythonCodes/src/communication/secure_protocol.py
import asyncio
import websockets
import json
from typing import Dict, Any, Optional
from cryptography.fernet import Fernet
from .quantum_encryption import QuantumEncryption
from ..utils.error_handling import KAPESITError, ErrorSeverity
from ..utils.config_manager import config_manager

class SecureProtocol:
    def __init__(self):
        self.quantum_enc = QuantumEncryption()
        self.session_keys: Dict[str, bytes] = {}
        self.fernet_keys: Dict[str, Fernet] = {}
        
    async def establish_secure_channel(self, websocket: websockets.WebSocketServerProtocol) -> str:
        try:
            # Generate quantum key
            quantum_key = self.quantum_enc.generate_quantum_key()
            
            # Create session key
            session_key = Fernet.generate_key()
            fernet = Fernet(session_key)
            
            # Encrypt session key with quantum key
            encrypted_key = self.quantum_enc.encrypt_message(session_key.decode(), quantum_key)
            
            # Send encrypted session key
            await websocket.send(json.dumps({
                "type": "key_exchange",
                "key": encrypted_key.hex()
            }))
            
            # Store session information
            session_id = websocket.id
            self.session_keys[session_id] = session_key
            self.fernet_keys[session_id] = fernet
            
            return session_id
            
        except Exception as e:
            raise KAPESITError(
                f"Secure channel establishment failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="SECURE_PROTOCOL"
            )
    
    async def encrypt_message(self, session_id: str, message: Dict[str, Any]) -> bytes:
        try:
            if session_id not in self.fernet_keys:
                raise ValueError("Invalid session ID")
                
            fernet = self.fernet_keys[session_id]
            message_str = json.dumps(message)
            return fernet.encrypt(message_str.encode())
            
        except Exception as e:
            raise KAPESITError(
                f"Message encryption failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="SECURE_PROTOCOL"
            )
    
    async def decrypt_message(self, session_id: str, encrypted: bytes) -> Dict[str, Any]:
        try:
            if session_id not in self.fernet_keys:
                raise ValueError("Invalid session ID")
                
            fernet = self.fernet_keys[session_id]
            decrypted = fernet.decrypt(encrypted)
            return json.loads(decrypted.decode())
            
        except Exception as e:
            raise KAPESITError(
                f"Message decryption failed: {str(e)}",
                severity=ErrorSeverity.ERROR,
                component="SECURE_PROTOCOL"
            )
    
    def close_session(self, session_id: str) -> None:
        if session_id in self.session_keys:
            del self.session_keys[session_id]
        if session_id in self.fernet_keys:
            del self.fernet_keys[session_id] 